app/controllers/sessions_controller.rb

```diff
@@ -27,10 +28,39 @@ def create
   def twitter
     # raise request.env["omniauth.auth"].to_yaml
     auth = request.env["omniauth.auth"]
-    user = User.where(uid: auth["uid"]).first || User.from_twitter(auth)
-    session[:user_id] = user.id
-    flash[:notice] = "You have been logged in through Twitter."
-    redirect_back_or root_url
+    if session[:user_id].nil?  #user is not currently logged in.
+      user = User.where(uid: auth["uid"]).first || User.from_twitter(auth)
+      session[:user_id] = user.id  #Note:  user.id & user.uid are different.  uid = twitter
+      flash[:notice] = "You have been logged in through Twitter."
+      redirect_back_or root_url
+    elsif User.where(uid: auth["uid"]).empty? || session[:user_id] == User.where(uid: auth["uid"]).first.id
+      #Either no Twitter account already existed OR
+      #user is logged in & the Twitter account was already linked to the current session id
+      #overwrite current user info with what is in Twitter account
+      user = User.where(id: session[:user_id]).first
+      user.name = auth.info.name
+      user.uid = auth.uid
+      user.provider = auth.provider
+      user.save
+      flash[:notice] = "Your account details have been synced with Twitter."
+      redirect_back_or root_url
+    else
+      #Twitter id exists, but it is not the current user
+      #merge data and merge saved links
+      #should add a pop up in the case that current account has a twitter asscn, but
+      #asscn is different from what user just authenticated
+
+      current_user = User.where(id: session[:user_id]).first
+      twitter_user = User.where(uid: auth["uid"]).first
+      current_user.name = twitter_user.name
+      current_user.uid = twitter_user.uid
+      current_user.provider = twitter_user.provider
+      current_user.save
+      twitter_user.links.each { |l| l.update user_id: current_user.id }
```

Comment:
I'm not 100% sure about all of this... Also, if the `twitter_user` has thousands of links, this action could take FOREVER.